### ISSUE TITLE
dllunix.so: undefined symbol: caml_ba_element_size

### ISSUE THREAD

**Issue by timburgess** [darioteixeira/pgocaml#52]
I've resolved this but I thought I would post this issue in the event that someone else hits the same issue:

I upgraded to `opam 2.0.0` recently and afterwards went to install `pgocaml` with `opam install pgocaml`. That failed and I then pinned to the repo for the latest source with
 `opam pin https://github.com/darioteixeira/pgocaml.git`

But installing produced:

```
<><> Synchronising pinned packages ><><><><><><><><><><><><><><><><><><><><><><>
[pgocaml.2.3+dev] no changes from git+https://github.com/darioteixeira/pgocaml.git

The following actions will be performed:
  ∗ install pgocaml 2.3+dev*

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of pgocaml failed at "/home/tim/.opam/opam-init/hooks/sandbox.sh build make".

#=== ERROR while compiling pgocaml.2.3+dev ====================================#
# context     2.0.0 | linux/x86_64 | base-bigarray.base base-threads.base base-unix.base ocaml-base-compiler.4.06.1 | pinned(git+https://github.com/darioteixeira/pgocaml.git#4ecb0c55)
# path        ~/.opam/4.06.1/.opam-switch/build/pgocaml.2.3+dev
# command     ~/.opam/opam-init/hooks/sandbox.sh build make
# exit-code   2
# env-file    ~/.opam/log/pgocaml-9205-a71bb4.env
# output-file ~/.opam/log/pgocaml-9205-a71bb4.out
### output ###
# [...]
# /home/tim/.opam/4.06.1/bin/ocamlfind ocamldep -package bytes -package calendar -syntax camlp4o -package camlp4.extend -package camlp4.quotations -package csv -package hex -package re -package unix -modules src/PGOCaml_aux.mli > src/PGOCaml_aux.mli.depends
# + /home/tim/.opam/4.06.1/bin/ocamlfind ocamldep -package bytes -package calendar -syntax camlp4o -package camlp4.extend -package camlp4.quotations -package csv -package hex -package re -package unix -modules src/PGOCaml_aux.mli > src/PGOCaml_aux.mli.depends
# Fatal error: cannot load shared library dllunix
# Reason: /home/tim/.opam/4.06.1/lib/ocaml/stublibs/dllunix.so: undefined symbol: caml_ba_element_size
# File "src/PGOCaml_aux.mli", line 1:
# Error: Error while running external preprocessor
# Command line: camlp4 -nolib '-I' '/home/tim/.opam/4.06.1/lib/camlp4' '-I' '/home/tim/.opam/4.06.1/lib/camlp4' '-I' '/home/tim/.opam/4.06.1/lib/camlp4' 'pa_o.cmo' 'pa_op.cmo' 'pr_dump.cmo' 'pa_extend.cmo' 'q_MLast.cmo'  'src/PGOCaml_aux.mli' > /tmp/ocamlpp2f890f
# 
# Command exited with code 2.
# E: Failure("Command ''/home/tim/.opam/4.06.1/bin/ocamlbuild' src/pgocaml.cma src/pgocaml.cmxa src/pgocaml.a src/pgocaml.cmxs utils/pgocaml_prof.byte -tag debug' terminated with error code 10")
# Makefile:7: recipe for target 'build' failed
# make: *** [build] Error 1
```
Looking at the details, the crux of this is 
`dllunix.so: undefined symbol: caml_ba_element_size` so I went googling.

`opam install camlp4` produced the required dependency and this time `opam install pgocaml` worked.

**Comment by jrochel** [comment]
@timburgess : Have a look at the discussion on https://github.com/ocaml/opam-repository/pull/13707. (You need to unfold the thread marked as "outdated".) I had encountered this issue when compiling pgocaml on Travis CI while publishing to opam.
@dra27 resolved it in https://github.com/ocaml/opam-repository/pull/13750. I didn't look at it though.
@dra27, maybe you can share your insights here?

**Comment by dra27** [comment]
I had initially thought that this was a problem with the vanilla release of opam 2.0 (which led to https://github.com/ocaml/opam-repository/pull/13733), but the issue I was thinking of was actually https://github.com/ocaml/opam/pull/3164 which was merged well before 2.0.0 was released.

The actual issue here is with a system-installed `camlp4` - your switch is OCaml 4.06.1, but I expect you have/had an older version of OCaml installed by your system package manager with an accompanying camlp4 (Ubuntu/Debian?).

There's a bug (or lack of detection) in `ocamlfind`'s configure in that it assumes that any `camlp4` it finds in `PATH` belongs to the compiler it also found in `PATH`, but this isn't true in your setup (the current compiler will be `/home/tim/.opam/4.06.1/bin/ocamlc`, but `camlp4` is probably `/usr/bin/camlp4`). IIRC `camlp4` in older releases is a bytecode executable, so it'll load up `/usr/bin/ocamlrun` (probably 4.02.3) but then attempt to load `dllunix.so` from the 4.06.1 switch.

The fix in opam-repository is to disable ocamlfind's detection of camlp4, so if you run `opam update` you shouldn't see this error again.

**Comment by jrochel** [comment]
Thanks.
